pkg/compose/run.go

```diff
@@ -115,22 +118,18 @@ func (s *composeService) runInteractive(ctx context.Context, containerID string,
 	}
 }
 
-func (s *composeService) terminateRun(ctx context.Context, containerID string, opts api.RunOptions, err error) (int, error) {
-	if err != nil {
-		return 0, err
-	}
-	inspect, err := s.apiClient.ContainerInspect(ctx, containerID)
-	if err != nil {
-		return 0, err
-	}
-	exitCode := 0
-	if inspect.State != nil {
-		exitCode = inspect.State.ExitCode
+func (s *composeService) terminateRun(ctx context.Context, containerID string, opts api.RunOptions) (exitCode int, err error) {
+	exitCh, errCh := s.apiClient.ContainerWait(ctx, containerID, container.WaitConditionNotRunning)
+	select {
+	case exit := <-exitCh:
+		exitCode = int(exit.StatusCode)
+	case err = <-errCh:
+		return
 	}
 	if opts.AutoRemove {
 		err = s.apiClient.ContainerRemove(ctx, containerID, moby.ContainerRemoveOptions{})
 	}
```

Comment:
Note that the `AutoRemove` option in API v1.30 and up (Docker 17.06 and up) is handled by the daemon (it will remove the container when it exits) (assuming the `AutoRemove` option is passed to the daemon when compose creates the container)

While it's of course possible for the client to also (manually) remove the container, there may be a race condition in that case (container no longer exists because it's already removed, or an error that the container is "already in the process of being removed").

See https://github.com/docker/cli/blob/135ffd205791001337d283310b286261242444e4/cli/command/container/run.go#L165-L183

And https://github.com/docker/cli/blob/135ffd205791001337d283310b286261242444e4/cli/command/container/utils.go#L16

Reply:
sure, I'd like we can just get rid of all this stuff by using some of the docker/cli `runXX` functions - will need to make those public